### PR TITLE
Add Store::isTrustedClient()

### PR DESCRIPTION
**Pull request by MatthewCroughan**
This PR extends the worker protocol in [`worker-protocol-hh`](https://github.com/NixOS/nix/blob/a4604f19284254ac98f19a13ff7c2216de7fe176/src/libstore/worker-protocol.hh) by adding a new operation (47) which permits a client to check if they are trusted. I have needed this for some scripts I'm trying to write which use Nix,  that need to check.

In addition, it also adds a new command `nix store check-trusted` which will make use of this addition. It will return exit code 0 if trusted, 1 if not trusted or 2 if an unexpected error occurs.

I have also added a new `checkInfo` function to [`doctor.cc`](https://github.com/NixOS/nix/blob/9a640afc1e63bed28926cb44fe85137fb7d2d2d1/src/nix/doctor.cc) which will print an inconsequential blue `[INFO]` telling the user whether they are trusted. This does not change the exit status of the `nix doctor` command.

image

The following Bash function is an example of a fallback mechanism I've implemented in Bash but do not have the skills to implement in C++.

```bash
function isTrustedUser {
  function writeDummyBuildLog {
    {
      echo 0x6378696e 0x00000000; # WORKER_MAGIC_1
      echo 0x0a010000 0x00000000; # PROTOCOL_VERSION
      echo 0x2d000000 0x00000000; # wopAddBuildlog (45)
      echo 0x26000000 0x00000000; # Length of Store Path (38)
      echo -n "$(echo -n aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-a.drv | xxd -p -c0)00000000000000000000"; # Store Path
    } \
    | xxd -p -r \
    | socat - UNIX-CONNECT:/nix/var/nix/daemon-socket/socket \
    | tr -d '\0'
  }
  x=$(writeDummyBuildLog)
  if (grep -aq "error:" <<< "$x") && (grep -aq "not privileged" <<< "$x")
  then
    echo "You are not a trusted user"
    return 1
  elif (grep -aq "error:" <<< "$x")
    echo "An unknown error occurred whilst checking if you are a trusted user"
    return 2
  else
    :
    return 0
  fi
}
isTrustedUser
```

Thanks a lot to @roberth who introduced me to the C++ codebase and answering a lot of questions. And thanks to @tomberek + @pkharvey + @helicalbytes for helping me understand the daemon protocol.

### CODE REVIEW

**Comment by MatthewCroughan** [comment]
CI fails because it is using the old daemon protocol version, this is an expected failure.

**Comment by NobbZ** [comment]
This is a good first step into a solution for #6752/#6672.

Can't say much codewise though.

**Comment by MatthewCroughan** [comment]
This PR will also improve https://github.com/NixOS/nix/pull/7461

File: src/nix/doctor.cc
Comment (max-privatevoid) [inline src/nix/doctor.cc:141]: This will throw an error if the remote store is too old, possibly prematurely terminating the doctor, which feels a little surprising. I wouldn't call it a dealbreaker though.

**Review by max-privatevoid — CHANGES_REQUESTED** [review]

File: src/libstore/dummy-store.cc
Comment (MatthewCroughan) [reply src/libstore/dummy-store.cc:62]: On the advice of @roberth I decided not to do this. If I was going to do this, I'd just make the default implementation in the `virtual` function in `store-api.cc` say "This store type is not supported" for example.

File: src/nix/doctor.cc
Comment (MatthewCroughan) [reply src/nix/doctor.cc:141]: It doesn't change the exit status, and I think it's the correct way to fail. Failure is not a form of `[INFO]`, only the successful case of figuring out if we're trusted or not is. Anything else should just be printed as-is, which is how it works at the moment.

Comment (MatthewCroughan) [reply src/nix/doctor.cc:141]: Apparently you need to bit shift by 8 bytes, this was incredibly confusing to me because:
- `GET_PROTOCOL_MAJOR(conn->daemonVersion)` will return `256`
- `GET_PROTOCOL_MAJOR(conn->daemonVersion) >> 8` will return `1`

Anyway.. I've implemented it now as follows.

```diff
diff --git a/src/libstore/remote-store.cc b/src/libstore/remote-store.cc
index 154afa726..5cfd432dd 100644
--- a/src/libstore/remote-store.cc
+++ b/src/libstore/remote-store.cc
@@ -1092,7 +1092,8 @@ bool RemoteStore::checkTrustedUser()
         conn.processStderr();
         return readInt(conn->from);
     } else {
-      throw Error("daemon protocol %d is too old (< 2.13) to ask whether we're a trusted user",
+      throw Error("daemon protocol %d.%d is too old (< 2.13) to ask whether we're a trusted user",
+          GET_PROTOCOL_MAJOR(conn->daemonVersion) >> 8,
           GET_PROTOCOL_MINOR(conn->daemonVersion));
     }
 }
```

File: src/libstore/remote-store.cc
Comment (MatthewCroughan) [reply src/libstore/remote-store.cc:1097]: Apparently you need to bit shift by 8 bytes to read `GET_PROTOCOL_MAJOR`. This was incredibly confusing to me because:
- `GET_PROTOCOL_MAJOR(conn->daemonVersion)` will return `256`
- `GET_PROTOCOL_MAJOR(conn->daemonVersion) >> 8` will return `1`

Anyway.. I've implemented it now as follows.

```diff
diff --git a/src/libstore/remote-store.cc b/src/libstore/remote-store.cc
index 154afa726..5cfd432dd 100644
--- a/src/libstore/remote-store.cc
+++ b/src/libstore/remote-store.cc
@@ -1092,7 +1092,8 @@ bool RemoteStore::checkTrustedUser()
         conn.processStderr();
         return readInt(conn->from);
     } else {
-      throw Error("daemon protocol %d is too old (< 2.13) to ask whether we're a trusted user",
+      throw Error("daemon protocol %d.%d is too old (< 2.13) to ask whether we're a trusted user",
+          GET_PROTOCOL_MAJOR(conn->daemonVersion) >> 8,
           GET_PROTOCOL_MINOR(conn->daemonVersion));
     }
 }
```

**Comment by MatthewCroughan** [comment]
I somehow accidentally removed the review request for @edolstra @fricklerhandwerk and @thufschmitt just by requesting a re-review from @max-privatevoid, this seems to be a bug in the way reviews are handled on Github.

File: src/libstore/dummy-store.cc
Comment (max-privatevoid) [reply src/libstore/dummy-store.cc:62]: A default implementation was what I had in mind, yes. My bad for being unclear.

Comment (MatthewCroughan) [reply src/libstore/dummy-store.cc:62]: @roberth seems not to like default implementations, and I also think it's a good idea to avoid default implementations, generally. It's a small nitpick and something that seems to vary between programmers.

**Comment by MatthewCroughan** [comment]
I marked the PR as a draft, then unmarked it as a draft. It seems that the GitHub review bug is still in effect.

File: src/libstore/remote-store.cc
Comment (roberth) [reply src/libstore/remote-store.cc:1097]: https://github.com/NixOS/nix/pull/7523

**Review by max-privatevoid — APPROVED** [review]

**Comment by MatthewCroughan** [comment]
Just pushed to resolve conflicts with the release notes.

File: src/libstore/daemon.cc
Comment (roberth) [inline src/libstore/daemon.cc:975]:
```suggestion
    case wopIsTrustedUser: {
        logger->startWork();
        bool result = trusted && store->isTrustedUser();
```

These two symbols are only about retrieving the information, but not necessarily acting upon it, so it's not really a check.
The `nix store check-trusted` command does not need renaming, because returning an exit code is more of a side effect.

File: src/libstore/build/local-derivation-goal.cc
Comment (roberth) [inline src/libstore/build/local-derivation-goal.cc:1470]: It may make sense to call recursive nix not trusted.

Comment (MatthewCroughan) [reply src/libstore/build/local-derivation-goal.cc:1470]: In which case, should we just return false?

File: src/libstore/daemon.cc
Comment (MatthewCroughan) [reply src/libstore/daemon.cc]: I've kept the local function name `checkTrustedUser()` in check-trusted.cc and doctor.cc, but otherwise each store implementation to `isTrustedUser`. Thanks for teaching me more about naming conventions!

File: src/libstore/store-api.hh
Comment (thufschmitt) [inline src/libstore/store-api.hh:639]: I think it's fine to define this to `unsupported` by default like we do for a bunch of other methods and remove all the custom implementations for the stores that don't support it:

```suggestion
    virtual bool isTrustedUser()
    { unsupported("isTrustedUser"); }
```

File: src/libstore/remote-store.cc
Comment (thufschmitt) [inline src/libstore/remote-store.cc:1097]: I actually don't think that printing the daemon protocol version is useful here since it's mostly an internal detail. It would be better to point to the minimal working Nix version (2.13 in that instance).

File: src/nix/doctor.cc
Comment (thufschmitt) [inline src/nix/doctor.cc:143]: Do we need to print the store uri here? It doesn't seem very useful (or if we want to print it, we could have it as its own line somewhere else)

File: src/nix/store-check-trusted.cc
Comment (thufschmitt) [inline src/nix/store-check-trusted.cc:61]:
```suggestion
static auto rCmdCheckTrusted = registerCommand2<CmdCheckTrusted>({"store", "check-trusted-user"});
```

`trusted` has a bunch of different meanings in Nix, better be explicit

File: src/nix/store-check-trusted.cc
Comment (thufschmitt) [inline src/nix/store-check-trusted.cc:39]: Better keep the description short and add some proper doc of how this works, something like:

```suggestion
    std::string description() override
    {
        return "check whether you are a trusted-user";
    }
    
    std::string doc() override
    {
          return                                                                                                                                     
              #include "store-check-trusted.md"                                                                                                                     
              ;    
    }
```

Where `store-check-trusted.md` would contain some more docs for the command (similar to https://github.com/NixOS/nix/blob/master/src/nix/ping-store.md for example)

File: src/nix/store-check-trusted.cc
Comment (thufschmitt) [inline src/nix/store-check-trusted.cc:44]:
```suggestion
        bool success = checkTrustedUser(store);
```

Unless there's a good reason to make it a class variable, I'd rather keep it local as it makes things much more readable

File: src/nix/store-check-trusted.cc
Comment (thufschmitt) [inline src/nix/store-check-trusted.cc:26]: I don't think these colors bring a lot here. This is low-level plumbing, we should probably stick to the most sober solution

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Thanks, that's a useful command to have I think.

Left some suggestions on the code and docs, but that looks good overall.

(That also makes me realize with horror that we don't have any test for the "non-trusted-user" case :grimacing: )

File: src/libstore/remote-store.cc
Comment (MatthewCroughan) [reply src/libstore/remote-store.cc:1097]: I also felt that this was unnecessary, but just copied the wisdom/pattern from other parts of the code. I'll remove it.

File: src/nix/doctor.cc
Comment (MatthewCroughan) [reply src/nix/doctor.cc:143]: I think it's useful for debug, especially in doctor. If you do `nix doctor` and my store is a remote store, I want to know. It helps me debug my configuration. Granted this could also be done by increasing Nix verbosity with  `-v`. My preference is to keep printing it.

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:26]: Then we should be removing that in another PR, for `nix doctor` itself. But for this PR to remain a simple addition, I'm not going to touch the existing code/pattern that doctor.cc implements.

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:44]: This is a pattern I copied from doctor.cc, nothing intentional.

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:39]: > proper doc of how this works

The proper doc of how this works is that it returns 1, 0 or 2. I don't think it's possible to make a dedicated document, unless you want it to describe the implementation details, which I think would not be suitable.

Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:39]: Are there other commands with similarly very short and possibly useless manual pages? I think this would actually hide the information? `nix doctor` does not have its own man page, for the same reason, it's too small to document.

File: src/nix/store-check-trusted.cc
Comment (thufschmitt) [reply src/nix/store-check-trusted.cc:26]: No, I think that's fine for `doctor` which is intended as a higher-level “check for obvious configuration issues” command. I just think we don't need it in this command

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:61]: What other meanings does `trusted` have in Nix? And is this not disambiguated already by the fact that it is a subcommand of `store`?

Comment (thufschmitt) [reply src/nix/store-check-trusted.cc:61]: The three that come to my mind are:

- A path is trusted if it has a valid signature or has been built locally
- A public key is trusted or not
- A substituter can also be marked as [trusted](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-trusted-substituters) 

None of these would _totally_ make sense here (except maybe the third one), but they are closely related to the store and I can imagine them being misleading for a user

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:26]: Ah, I misread the source file name, I thought the code highlighted was from doctor.cc

In this case, it's not displayed unless your input `isatty`. This means it handles the case of scripting where you don't want any stdout, only an exit code, as well as also telling the user whether they passed or failed trust checking.

It is only in the interactive case where the input is a tty, where you would want any output from the command at all. So what would you rather it print? The same text without the pass/fail context text prepended?

File: src/nix/store-check-trusted.cc
Comment (thufschmitt) [reply src/nix/store-check-trusted.cc:39]: > The proper doc of how this works is that it returns 1, 0 or 2. I don't think it's possible to make a dedicated document, unless you want it to describe the implementation details, which I think would not be suitable.

Fair, a separate document might be overkill. But keeping the description short and moving the exact behaviour separate (even a simple one like here) is important for things like `nix --help` (which will list all the commands with their description).

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:61]: That makes sense, I'll change it.

File: src/libstore/store-api.hh
Comment (MatthewCroughan) [reply src/libstore/store-api.hh:639]: That definitely makes it a smaller diff too.

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:26]: After some thinking, I think I agree now. I was just copying Nix doctor and liked the pretty colors. But it's not sober, as you mention.

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc:39]: After breaking it out, I found that showing an example with `--store` is beneficial anyway and have created the md file.

File: src/nix/store-check-trusted.cc
Comment (MatthewCroughan) [reply src/nix/store-check-trusted.cc]: I've removed the color and fail/pass text.

File: src/libstore/store-api.hh
Comment (MatthewCroughan) [reply src/libstore/store-api.hh]: Done. I've left the implementation for legacy ssh though, since it's more helpful to print it.

**Comment by MatthewCroughan** [comment]
@thufschmitt I believe I've adjusted the code to meet all of your suggestions, except for `nix doctor` printing the Store URI, which I think is quite helpful. I've re-requested a review from you on GitHub, sadly this has the unintended side-effect of removing @edolstra due to what seems like either repo misconfiguration or a GitHub bug. Could you please add him back as a reviewer if he's required to get this merged, after you do your next review? Thank you!

File: src/libstore/build/local-derivation-goal.cc
Comment (MatthewCroughan) [reply src/libstore/build/local-derivation-goal.cc]: I've taken the suggestion from @thufschmitt to include a default implementation of `unsupported("foo")` instead. If we should return false, that is something we can do in another PR, if this gets merged soon.

**Comment by roberth** [comment]
> add him back as a reviewer if he's required to get this merged

This isn't required if we follow the Nix team "project board" process that we've started to work with. Technically we should approve the idea first, but 3/5 Nix team members have already been involved, so we'll probably move this to final review and assign a reviewer in the next meeting.

Question for the team: is the process always required?

File: src/libstore/build/local-derivation-goal.cc
Comment (roberth) [reply src/libstore/build/local-derivation-goal.cc]: > just return false?

Yes.

Comment (MatthewCroughan) [reply src/libstore/build/local-derivation-goal.cc]: I've returned `false` for the case of recursive nix now.

File: src/libstore/store-api.hh
Comment (thufschmitt) [reply src/libstore/store-api.hh]: @roberth since the default behaviour is to fail loudly, I think it's OK because it's a conservative approach, and we can always refine the implementation after the fact if we realize that we've been missing one

File: doc/manual/src/release-notes/rl-next.md
Comment (thufschmitt) [inline doc/manual/src/release-notes/rl-next.md:29]: Merge failure? :upside_down_face:

File: doc/manual/src/release-notes/rl-next.md
Comment (thufschmitt) [inline doc/manual/src/release-notes/rl-next.md:37]: I find this point a bit hard to read (and technically misleading since `nix doctor` doesn't really call `nix store check-trusted-user`). What about simply:

```suggestion

* `nix doctor` now displays whether you are a trusted user or not.
```

**Review by thufschmitt — CHANGES_REQUESTED** [review]
Small glitch in the release notes, but that's looking good otherwise :+1: 

Question for the crowd (@NixOS/nix-team): Should we follow #5930 and feature-gate this for some time?

**Comment by fricklerhandwerk** [comment]
@roberth this needs discussion by the team and is out of scope here. Opened https://github.com/NixOS/nix/pull/7579 for that purpose.
@thufschmitt I'd be in favor of making it experimental, for the exact reason stated in the relevant PR. It would give us at least some escape hatch if a substantial change is needed. For this specific feature, while I'm on board with the use case it unblocks, I'd still like to be cautious with growing the API surface.

<details><summary>Side notes</summary>

Both cases constitute additional [red tape](https://en.m.wikipedia.org/wiki/Red_tape). We want to make the contribution process easier, but also be much more deliberate about the overall design. Right now, due to many design decisions and strategies lacking, these may appear as conflicting goals; we already had multiple discussions over adding features where the use case was deemed worthy of support, but the specific change seemed too risky in terms of keeping options for the future. As a result, progress on these issues is stalling. Examples: https://github.com/NixOS/nix/pull/7188#issuecomment-1377200466, https://github.com/NixOS/nix/pull/545#issuecomment-1377208634.

I think the team should strive towards providing clear (written) guidance on how to approach new feature development in order to make it more likely to succeed.

</details>

File: doc/manual/src/release-notes/rl-next.md
Comment (MatthewCroughan) [reply doc/manual/src/release-notes/rl-next.md:29]: Derp

File: doc/manual/src/release-notes/rl-next.md
Comment (MatthewCroughan) [reply doc/manual/src/release-notes/rl-next.md]: I like that more. Fixed.

File: doc/manual/src/release-notes/rl-next.md
Comment (MatthewCroughan) [reply doc/manual/src/release-notes/rl-next.md]: Fixed

**Comment by MatthewCroughan** [comment]
I've rebased this PR on 2.14 after the Nix 2.13 release yesterday.

File: tests/remote-store.sh
Comment (Ericson2314) [inline tests/remote-store.sh:14]: This can be updated, but fine if you don't bother fixing this until we are about to merge so you aren't stuck doing it again.

Comment (MatthewCroughan) [reply tests/remote-store.sh:14]: This means the daemon needs to be newer than 2.13, so it is the correct value. When 2.13 was released, this was bumped from 2.12 to 2.13.

**Comment by Ericson2314** [comment]
@MatthewCroughan Do note with https://github.com/NixOS/nix/pull/7723 merged the `nix store ping` changes will be less work!

**Comment by MatthewCroughan** [comment]
Have dropped the ball on this temporarily, I'm a bit ill after FOSDEM, will pick this up when I am better, anyone else feel free to contribute by making a PR to this PR!

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting 2023-04-03:

- @Ericson2314 rewrote the PR based on the team's requested changes, needs a different reviewer
- assigned to @thufschmitt

**Comment by domenkozar** [comment]
Would be great to have this in for 2.15.

File: doc/manual/src/release-notes/rl-next.md
Comment (thufschmitt) [inline doc/manual/src/release-notes/rl-next.md:44]: This message feels a bit weird. Nix obviously already checked that. There just wasn't a way to get this information from the client side

File: src/libstore/remote-store.cc
Comment (thufschmitt) [inline src/libstore/remote-store.cc:236]: Let's not try to be confusingly smart here

```suggestion
                case 1:
                    conn.remoteTrustsUs = NotTrusted;
                    break;
                case 2:
                    conn.remoteTrustsUs = Trusted;
```

File: src/libstore/daemon.cc
Comment (thufschmitt) [inline src/libstore/daemon.cc:1041]: :exploding_head: 

Can we rather have a proper well-defined serialization scheme for `std::optional<TrustedFlag>` (in `worker-protocol.hh` like the rest I would assume)? This inlined arithmetic trick is too smart for its own good I think

File: src/libstore/remote-store.cc
Comment (thufschmitt) [reply src/libstore/remote-store.cc:236]: (also, I'd rather move this to a proper `worker_proto::read(..., TrustedFlag)` function like the rest)

**Review by thufschmitt — CHANGES_REQUESTED** [review]
The change is good overall, but the wire encoding/decoding logic is a bit too smart imho. Let's keep things simple as much as we can.

**Comment by Ericson2314** [comment]
Good point, I did sort of forget about that mechanism! :flushed:

File: src/libstore/remote-store.cc
Comment (thufschmitt) [inline src/libstore/remote-store.cc:56]: Saving it for a potential follow-up (and fairly low-prio), but it could be nicer to just warn and return `std::nullopt` in that case. Just to make potential future evolutions of the protocol nicer.

**Review by thufschmitt — APPROVED** [review]
Yay!

Just a small suggestion, but fairly minor and not blocking. So let's merge

File: src/libstore/remote-store.cc
Comment (Ericson2314) [reply src/libstore/remote-store.cc:56]: I had thought about that, but I think it's just better to bunl the protocol version on that case? Maybe I'm forgetting, but I couldn't think of another time we tried to do this sort of week *forward* compat before.

Comment (roberth) [reply src/libstore/remote-store.cc:56]: We currently have more or less one forward compat mechanism, which is the protocol version handshake.
If we're going to improve the forward compat mechanism, it would make more sense to do it structurally, with a self-describing or schema based serialization where we can add fields as we please.

**Pull request merged**
